scp/scpGenOperatorInterpreter.cpp

```diff
@@ -1,69 +1,110 @@
 /*
-* This source file is part of an OSTIS project. For the latest info, see http://ostis.net
-* Distributed under the MIT License
-* (See accompanying file COPYING.MIT or copy at http://opensource.org/licenses/MIT)
-*/
+ * This source file is part of an OSTIS project. For the latest info, see http://ostis.net
+ * Distributed under the MIT License
+ * (See accompanying file COPYING.MIT or copy at http://opensource.org/licenses/MIT)
+ */
 
+#include "scpGenOperatorInterpreter.hpp"
 #include "scpKeynodes.hpp"
 #include "scpUtils.hpp"
-#include "scpGenOperatorInterpreter.hpp"
+
 #include "gen_operators/SCPOperatorGenEl.hpp"
 #include "gen_operators/SCPOperatorGenElStr3.hpp"
 #include "gen_operators/SCPOperatorGenElStr5.hpp"
 #include "gen_operators/SCPOperatorSysGen.hpp"
-#include "sc-memory/sc_memory.hpp"
+
 #include <iostream>
 
-namespace scp {
-ScAddr ASCPGenOperatorInterpreter::msAgentKeynode;
+namespace scp
+{
+ScAddrToValueUnorderedMap<std::function<SCPOperator *(ScMemoryContext &, ScAddr)>>
+    ASCPGenOperatorInterpreter::supportedOperators = {};
 
-SC_AGENT_IMPLEMENTATION(ASCPGenOperatorInterpreter)
+ScResult ASCPGenOperatorInterpreter::DoProgram(
+    ScEventAfterGenerateOutgoingArc<ScType::EdgeAccessConstPosPerm> const & event,
+    ScAction & action)
 {
-    if (!edgeAddr.IsValid())
-        return SC_RESULT_ERROR;
-
-    ScAddr scp_operator =m_memoryCtx.GetEdgeTarget(edgeAddr);
-
-    ScAddr type;
-    if (SC_TRUE != Utils::resolveOperatorType(m_memoryCtx, scp_operator, type))
-        return SC_RESULT_ERROR_INVALID_TYPE;
-
-    SCPOperator* oper = nullptr;
-    if (type == Keynodes::op_genEl)
-    {
-        oper = new SCPOperatorGenEl(m_memoryCtx, scp_operator);
-    }
-    if (type == Keynodes::op_genElStr3)
-    {
-        oper = new SCPOperatorGenElStr3(m_memoryCtx, scp_operator);
-    }
-    if (type == Keynodes::op_genElStr5)
-    {
-        oper = new SCPOperatorGenElStr5(m_memoryCtx, scp_operator);
-    }
-    if (type == Keynodes::op_sys_gen)
-    {
-        oper = new SCPOperatorSysGen(m_memoryCtx, scp_operator);
-    }
-    if (oper == nullptr)
-        return SC_RESULT_ERROR_INVALID_PARAMS;
+  if (!event.GetArc().IsValid())
+    return action.FinishUnsuccessfully();
+
+  ScAddr scp_operator = event.GetOtherElement();
+
+  ScAddr type;
+  if (!Utils::resolveOperatorType(m_context, scp_operator, type))
+    return action.FinishUnsuccessfully();
+
+  SCPOperator * oper = nullptr;
+  auto const & pair = supportedOperators.find(type);
+  if (pair != supportedOperators.cend())
+    oper = pair->second(m_context, scp_operator);
+
+  if (oper == nullptr)
+    return action.FinishUnsuccessfully();
```

Comment:
Can you log error here for debug?

Reply:
Check everywhere, please

Reply:
I added LOG_ERROR because this line should never be executed, CheckInitiationCondition ensures that operator will be created